### DISCUSSION TITLE
I'm stupid, or the tool sucks?

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#221] [General]
Followed the guide to great a portable runtime.

`obb -src_file test.obe -to_dir test -to_name test`

Well, always said `>>> Invalid Objeck install directory: '' <<<`.

I have to feed the Objeck directory to it:

`obb -src_file test.obe -to_dir test -to_name test -install C:\objeck-lang`

```
>>> Error encounented build native runtime environment <<<
        filesystem error: cannot create directory: No such file or directory [test\test]
```

So, it doesn't automatically create the `test` directory for me.

I decided to create that directory manually. Then:

```
>>> Error encounented build native runtime environment <<<
        filesystem error: cannot copy: File exists [C:\objeck-lang\bin] [C:\test\test\runtime\bin]
```

I started to lost patient. I run the command again:

`>>> Target directory 'test\test' already exists <<<`

I decided to stop trying.

**Comment by objeck** [comment]
Odd, I will look at this tomorrow. Are you running under MSYS2? If so, what shell?

**Reply by ghost** [reply]
No, I'm running this on Windows Command Prompt.

**Comment by objeck** [comment]
How did you build the code? The [utility](https://github.com/objeck/objeck-lang/blob/master/core/utils/launcher/builder.cpp) is pretty simple. However, it uses newer C++ file and directory APIs. It worked when I tried it yesterday; however, the app was created at the directory root.

**Reply by ghost** [reply]
I confirm using your release it worked. I still have to specify Objeck directory via `-install`.

**Reply by objeck** [reply]
The issue is MSYS2's support for the C++ 17 ``filesystem`` APIs. The APIs are there; however, they work differently on MSYS2 and, I suspect, on Cygwin than on targets. MSYS2 does not support some of the launcher's APIs; however, I may be able to use their POSIX alternatives as a workaround.

**Reply by objeck** [reply]
@iahung thanks made ``-install`` a required parameter. At the time of development, I wanted to derive the location of the installation, as other sections of code do, must have given up or forgotten.

**Reply by ghost** [reply]
@objeck What about adding the environment variable OBJECK_HOME like Java's JAVA_HOME?

**Reply by ghost** [reply]
Btw, please tell me more about OBJECK_LIB_PATH. Is it a list of locations to search for libraries like Java's CLASSPATH or it's only allowed to contain only one location which is the directory objeck-lang/lib?

**Reply by objeck** [reply]
> Btw, please tell me more about OBJECK_LIB_PATH. Is it a list of locations to search for libraries like Java's CLASSPATH or it's only allowed to contain only one location which is the directory objeck-lang/lib?

It's is the location of .obl library files used by the compiler. It is also the location offset for native share libraries (.so, .dll) used at runtime.

**Reply by ghost** [reply]
So I'm forced to put .obl on that directory in order to have the compiler able to see it? On Java, they have the `CLASSPATH` environment variable (a list of locations to search for libraries). There is nothing equivalent on Objeck?

p/s: The same is also true for native libraries? I have to put them on `%OBJECK_LIB_PATH%\native` in order to have the compiler and the virtual machine to be able to see them?

**Reply by ghost** [reply]
Btw, what is the meaning of `-D_OBJECK_NATIVE_LIB_PATH` on makefiles?

**Reply by ghost** [reply]
Well, if there is an environment variable that contains a list of locations to search for libraries like Java's CLASSPATH it's much better. The locations of native libraries will be inferred based on the value of that environment variables.

For example, `OBJECK_CLASSPATH=C:\objeck-lang\lib;C:\working\objeck\lib` then the locations of native libraries will be inferred as `C:\objeck-lang\lib\native;C:\working\objeck\lib\native`.

**Reply by objeck** [reply]
> So I'm forced to put .obl on that directory in order to have the compiler able to see it? On Java, they have the `CLASSPATH` environment variable (a list of locations to search for libraries). There is nothing equivalent on Objeck?
> 
> p/s: The same is also true for native libraries? I have to put them on `%OBJECK_LIB_PATH%\native` in order to have the compiler and the virtual machine to be able to see them?

The idea was to have all external libraries loaded from known directories for security. Last ``%OBJECK_LIB_PATH%`` can reside anywhere.

**Reply by ghost** [reply]
But, why do we have to add `C:\objeck-lang\lib` to `OBJECK_LIB_PATH` anyway? Shouldn't it be searched by default? It could be inferred from the location of the running Objeck binary, e.g: `C:\objeck-lang\bin\..\lib`.

Or better yet, adding the environment variable `OBJECK_HOME` and infer from it.

**Reply by ghost** [reply]
@objeck Finally `OBJECK_LIB_PATH` is a list of locations like Java's `CLASSPATH` or it can contain only one location?

**Comment by objeck** [comment]
The ``-D_OBJECK_NATIVE_LIB_PATH`` preprocessor enables the use of the ``OBJECK_NATIVE_LIB_PATH`` variable. If it is not enabled, the compiler and runtime assume the ``lib`` directory is relative to the ``bin`` directory. One may specify a different location for ``OBJECK_NATIVE_LIB_PATH`` but only one location can be specified.
